File: 2022/day8/day8.py
import logging

logging.basicConfig(level=logging.DEBUG)


def count_left(mat, i, j):
    k = 1
    for k in range(1, j + 1):
        if mat[i][j - k] >= mat[i][j]:
            break
    return k


def count_right(mat, i, j):
    k = 1
    n_cols = len(mat[0]) - 1
    for k in range(1, n_cols - j + 1):
        if mat[i][j + k] >= mat[i][j]:
            break
    return k


def count_up(mat, i, j):
    k = 1
    for k in range(1, i + 1):
        if mat[i - k][j] >= mat[i][j]:
            break
    return k


def count_down(mat, i, j):
    k = 1
    n_rows = len(mat)
    for k in range(1, n_rows - i):
        if mat[i + k][j] >= mat[i][j]:
            break
    return k


def visible_from_left(mat, i, j):
    for k in range(0, j):
        if mat[i][k] >= mat[i][j]:
            return False
    return True


def visible_from_right(mat, i, j):
    for k in range(j + 1, len(mat[0])):
        if mat[i][k] >= mat[i][j]:
            return False
    return True


def visible_from_top(mat, i, j):
    for k in range(0, i):
        if mat[k][j] >= mat[i][j]:
            return False
    return True


def visible_from_bottom(mat, i, j):
    for k in range(i + 1, len(mat)):
        if mat[k][j] >= mat[i][j]:
            return False
    return True


def is_visible(mat, i, j):
    return (
        visible_from_left(mat, i, j)
        or visible_from_right(mat, i, j)
        or visible_from_top(mat, i, j)
        or visible_from_bottom(mat, i, j)
    )


def read(fname):
    with open(fname, "r") as f:
        mat = f.readlines()
    mat = [[int(i) for i in line.strip()] for line in mat]
    return mat


def scenic_score(mat, i, j):
    # first row or column
    if i * j == 0:
        return 0
    # last row or column
    if i == len(mat) - 1 or j == len(mat[0]) - 1:
        return 0
    return (
        count_left(mat, i, j)
        * count_right(mat, i, j)
        * count_up(mat, i, j)
        * count_down(mat, i, j)
    )


def max_scenic_score(mat):
    max_score = 0
    for i in range(1, len(mat) - 1):
        for j in range(1, len(mat[0]) - 1):
            score = scenic_score(mat, i, j)
            if score > max_score:
                max_score = score
                logging.debug("Found new max score of %d at (%d, %d)", max_score, i, j)
    return max_score


if __name__ == "__main__":
    mat = read("sample.txt")
    assert is_visible(mat, 1, 2)
    for i in range(5):
        assert is_visible(mat, i, 0)
        assert is_visible(mat, 0, i)
    n = sum(is_visible(mat, i, j) for i in range(len(mat)) for j in range(len(mat[0])))
    assert n == 21, n

    # part 2
    assert count_left(mat, 1, 2) == 1
    assert count_left(mat, 0, 3) == 3
    assert count_right(mat, 1, 2) == 2
    assert count_up(mat, 1, 2) == 1
    assert count_down(mat, 1, 2) == 2
    assert scenic_score(mat, 1, 2) == 4
    assert count_left(mat, 3, 2) == 2
    assert count_right(mat, 3, 2) == 2
    assert count_up(mat, 3, 2) == 2
    assert count_down(mat, 3, 2) == 1
    assert max_scenic_score(mat) == 8, max_scenic_score(mat)

    mat1 = read("input.txt")
    n = sum(
        is_visible(mat1, i, j) for i in range(len(mat1)) for j in range(len(mat1[0]))
    )
    logging.info("part1 = %d", n)

    logging.info("part2 = %d", max_scenic_score(mat1))

File: 2022/day8/day8.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func visibleFromLeft(X [][]int, row int, col int) bool {
	// always visible from edge
	if row*col == 0 {
		return true
	}
	for j := col - 1; j >= 0; j-- {
		if X[row][j] >= X[row][col] {
			return false
		}
	}
	return true
}
func visibleFromRight(X [][]int, row int, col int) bool {
	for j := col + 1; j < len(X); j++ {
		if X[row][j] >= X[row][col] {
			return false
		}
	}
	return true
}
func visibleFromTop(X [][]int, row int, col int) bool {
	for i := 0; i < row; i++ {
		if X[i][col] >= X[row][col] {
			return false
		}
	}
	return true
}
func visibleFromBottom(X [][]int, row int, col int) bool {
	for i := row + 1; i < len(X); i++ {
		if X[i][col] >= X[row][col] {
			return false
		}
	}
	return true
}

func isVisible(X [][]int, row int, col int) int {
	if visibleFromBottom(X, row, col) || visibleFromLeft(X, row, col) || visibleFromTop(X, row, col) || visibleFromRight(X, row, col) {
		return 1
	} else {
		return 0
	}
}

func countLeft(X [][]int, row int, col int) int {
	var k int
	for k = 1; col-k > 0 && X[row][col-k] < X[row][col]; k++ {
	}
	return k
}

func countRight(X [][]int, row int, col int) int {
	var k int
	for k = 1; col+k < len(X)-1 && X[row][col+k] < X[row][col]; k++ {
	}
	return k
}

func countUp(X [][]int, row int, col int) int {
	var k int
	for k = 1; row-k > 0 && X[row-k][col] < X[row][col]; k++ {
	}
	return k
}

func countDown(X [][]int, row int, col int) int {
	var k int
	for k = 1; row+k < len(X)-1 && X[row+k][col] < X[row][col]; k++ {
	}
	return k
}

func scenicScore(X [][]int, row int, col int) int {
	left := countLeft(X, row, col)
	right := countRight(X, row, col)
	up := countUp(X, row, col)
	down := countDown(X, row, col)
	fmt.Println(row, col, left, up, right, down)
	return up * left * right * down
}

func totalVisible(X [][]int) int {
	total := 0
	N := len(X[0])
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			total += isVisible(X, i, j)
		}
	}
	return total
}

func buildMatrix(filename string) [][]int {
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	matrix := make([][]int, 0)
	for scanner.Scan() {
		txt := strings.Split(scanner.Text(), "")
		row := make([]int, len(txt))
		for i, val := range txt {
			row[i], _ = strconv.Atoi(val)
		}
		matrix = append(matrix, row)
	}
	return matrix
}

func part1(filename string) int {
	matrix := buildMatrix(filename)
	return totalVisible(matrix)
}

func part2(filename string) int {
	matrix := buildMatrix(filename)
	maxScenicScore := -1
	for row := 1; row < len(matrix)-1; row++ {
		for col := 1; col < len(matrix)-1; col++ {
			score := scenicScore(matrix, row, col)
			if score > maxScenicScore {
				maxScenicScore = score
			}
		}

	}
	return maxScenicScore
}

func main() {
	var answer int
	answer = part1("sample.txt")
	if answer != 21 {
		panic("wrong answer to sample problem")
	}
	fmt.Println("the answer to part 1 is ")
	fmt.Println(part1("input.txt"))

	answer = part2("sample.txt")
	fmt.Println(answer)
	fmt.Println("the answer to part 2 is ")
	fmt.Println(part2("input.txt"))
}
